AIAgent/ml/validation/coverage/game_managers/model/process_game_manager.py

```diff
@@ -0,0 +1,387 @@
+                        f"Not all steps exhausted on {game_map.MapName} with non-100% coverage. Steps taken by oracle: {steps_count}, actual coverage: {actual_coverage_formatted}."
+                    )
+                logging.info(
+                    f"Finished map {map_name} "
+                    f"in {result.steps_count} steps,"
+                    f"actual coverage: {result.actual_coverage_percent:.2f}"
+                )
+                return result
+            except (FileNotFoundError, ValueError) as e:
+                ret_code = proc.poll()
+                if ret_code is None:
+                    time.sleep(3)
+                    continue
+                if isinstance(e, ValueError):
+                    msg = f"Incorrect result of {map_name}"
+                else:
+                    msg = f"The result of {map_name} cannot be found after completing the game?!"
+                logging.error(msg)
+                return GameFailed(msg)
+
+    def get_game_steps(self, game_map: GameMap) -> Optional[list[HeteroData]]:
+        game_map_info = self._games_info.get(game_map.MapName)
+        return (
+            game_map_info.total_steps
+            if game_map_info and game_map_info.total_steps
+            else None
+        )
+
+    def notify_steps_requirement(self, game_map: GameMap, required: bool):
+        map_name = game_map.MapName
+        if map_name not in self._games_info:
+            msg = f"Can't find game info of {game_map.MapName}"
+            logging.warning(msg)
+            return
+        game_map_info = self._games_info[map_name]
+        if isinstance(game_map_info.game_result, GameResultDetails):
+            game_result = game_map_info.game_result
+        else:
+            self._get_and_log_proc_output(game_map_info.proc, logging.error)
+            return
+
+        def get_conn():
+            svm_connection_info = game_result.svm_connection_info
+            port = svm_connection_info.occupied_port
+            logging.debug(f"Server listening on port {port}")
+            conn, addr = svm_connection_info.socket.accept()
+            logging.debug(f"Connection from {addr}")
+            return conn
+
+        def alert_svm_about_step_saving(conn: socket.socket, need_to_save_steps: bool):
+            conn.sendall(bytes([1 if need_to_save_steps else 0]))
+            conn.shutdown(socket.SHUT_WR)
+
+        conn = get_conn()
+        alert_svm_about_step_saving(conn, required)
+
+        if (
+            not game_map_info.total_steps and required
+        ):  # there is no already taken steps
+
+            def get_steps_from_svm() -> list:
+                output_dir = self._get_output_dir(game_map)
+                steps = output_dir / f"{map_name}_steps"
+                game_map_info.proc.wait()
+                with open(steps, "br") as f:
+                    steps_json = json.load(f)
+                steps = list(map(lambda v: ModelGameStep.from_dict(v), steps_json))  # type: ignore
```

Comment:
Such tricks with types hint at bad design.

Reply:
Try this: https://github.com/lidatong/dataclasses-json?tab=readme-ov-file#use-my-dataclass-with-json-arrays-or-objects

<img width="752" alt="image" src="https://github.com/user-attachments/assets/1bd1ca1d-89e1-4482-98e2-fddb17e46ef9" />